runtime/error.c

```diff
@@ -1,12 +1,103 @@
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+
 #include "balrt.h"
+#include "third-party/libbacktrace/backtrace.h"
+
+#define INITIAL_PC_COUNT 1
+#define MAX_PC_COUNT 256
+#define SKIP_FROM_END 1 // This number counts from the last stack frame
+#define THREAD 0
+
+struct backtrace_state *state;
+
+typedef struct {
+    uint32_t npcs;
+    uint32_t szpcs;
+    PC *pcs;
+} Trace;
+
+static void onError(void *data, const char *msg, int errnum) {
+    fprintf(stdout, "Error : libbacktrace : %s\n", msg);
+    fflush(stdout);
+}
+
+static int onFrame(void *data, PC pc, const char *filename, int lineno, const char *function) {
+    if (function == NULL || filename == NULL)
+        return 0;
+
+    fprintf(stdout, "%s", function);
+    fprintf(stdout, " %s", filename);
+    fputs(":", stdout);
+    fprintf(stdout, "%" PRIi32, lineno);
+    putc('\n', stdout);
+    fflush(stdout);
+    return 0;
+}
+
+static int onPC(void *data, PC pc) {
+    Trace *trace = (Trace *)data;
+    uint32_t npcs = trace->npcs;
+    uint32_t szpcs = trace->szpcs;
+    if (unlikely(npcs == MAX_PC_COUNT))
+        return 1;
+    if (npcs == szpcs) {
+        szpcs = szpcs << 1;
+        trace->szpcs = szpcs;
+        void *p = realloc(trace->pcs, sizeof(PC) * szpcs);
+        if (p == 0)
+            return 1;
+        trace->pcs = (PC *)p;
+    }
+
+    PC *pcs = trace->pcs;
+    *(pcs + npcs) = pc;
+
+    trace->npcs = npcs + 1;
+    return 0;
+}
+
+static int getPCs(Trace *trace) {
+    state = backtrace_create_state(NULL, THREAD, onError, NULL);
+
+    void *p = malloc(sizeof(PC) * INITIAL_PC_COUNT);
+    if (p == 0)
+        return 1;
+    trace->pcs = (PC *)p;
+    return backtrace_simple(state, SKIP_FROM_END, onPC, onError, trace);
+}
 
 TaggedPtr _bal_error_construct(TaggedPtr message, int64_t lineNumber) {
-    ErrorPtr ep = _bal_alloc(sizeof(struct Error));
+    Trace trace = {0, INITIAL_PC_COUNT};
+    int err = getPCs(&trace);
+    if (err) {
+        fprintf(stdout, "%s\n", "Error occured when getting backtrace for error value");
```

Comment:
No. Recover from errors. Do not print stuff to stderr.

Reply:
Is it okay to print a message to stdout if an error occurred?

Reply:
I don't think so, one option is to keep a flag in `struct Error` and print it when you are printing the error.
I think we'd have a better idea if we knew what kind of cases would give us error here.

Reply:
Yes, we know cases where an error occurs.
1. Memory allocation/reallocation
2. Create backtrace state
3. Exceed MAX_PC_COUNT limit

Reply:
For 1, 3 we at least have a palatially correct stacktrace, maybe even in 2 (when does this happens?). I think we should keep an `internalErrorCode` in `struct Error` and set when creating.
When printing we can consider this, and do something appropriate.
eg:
```
    myfunc1(file1.bal:2)
    myfunc3(file1.bal:6)
    ... (some lines omitted due to exceeding maximum stack trace limit)
```

(maybe this pr is not the correct place to do the full implementation, but avoid printing during construction)

Reply:
I will change this. The 2nd case happens if an error occurs in `backtrace_create_state`

Reply:
Printing a backtrace should be on a best efforts basis. Print out as much of a backtrace as you can (up to MAX_PC_COUNT lines). You can print `...` if you are omitting some.

Reply:
Okay.
Also, we can do the same thing(set a flag if an error occurs) for error_callback functions without printing the error inside the error_callback function.

Reply:
Changed.